x-pack/plugins/ingest_manager/server/collectors/helpers.ts

```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { CoreSetup, SavedObjectsClient } from 'kibana/server';
+
+export async function getInternalSavedObjectsClient(core: CoreSetup) {
+  return core.getStartServices().then(async ([coreStart]) => {
+    // TODO: is this safe?
+    return (coreStart.savedObjects.createInternalRepository() as unknown) as SavedObjectsClient;
```

Comment:
Hi! Butting in here just to say I'm curious about this ... is there a reason you need this "internal repository" vs just `savedObjects.client` ? And if you do, what's does `createInternalRepository()` return before you re-cast it as unknown?

Reply:
I was working from this example code:
```
// server/plugin.ts
import { UsageCollectionSetup } from 'src/plugins/usage_collection/server';
import { CoreSetup, CoreStart } from 'kibana/server';

class Plugin {
  private savedObjectsRepository?: ISavedObjectsRepository;

  public setup(core: CoreSetup, plugins: { usageCollection?: UsageCollectionSetup }) {
    registerMyPluginUsageCollector(() => this.savedObjectsRepository, plugins.usageCollection);
  }

  public start(core: CoreStart) {
    this.savedObjectsRepository = core.savedObjects.createInternalRepository();
  }
}
```
(from https://github.com/elastic/kibana/blob/master/src/plugins/usage_collection/README.md#new-platform ) and poked at it until it started working.

`createInternalRepository()` returns a `SavedObjectsRepository`.

Trying to use it in place of a `SavedObjectClient` results in this type mismatch:

```
const soClient: Pick<SavedObjectsRepository, "create" | "bulkCreate" | "delete" | "deleteByNamespace" | "find" | "bulkGet" | "get" | "update" | "addToNamespaces" | "deleteFromNamespaces" | "bulkUpdate" | "incrementCounter">
Argument of type 'Pick<SavedObjectsRepository, "create" | "bulkCreate" | "delete" | "deleteByNamespace" | "find" | "bulkGet" | "get" | "update" | "addToNamespaces" | "deleteFromNamespaces" | "bulkUpdate" | "incrementCounter">' is not assignable to parameter of type 'SavedObjectsClient'.
  Type 'Pick<SavedObjectsRepository, "create" | "bulkCreate" | "delete" | "deleteByNamespace" | "find" | "bulkGet" | "get" | "update" | "addToNamespaces" | "deleteFromNamespaces" | "bulkUpdate" | "incrementCounter">' is missing the following properties from type 'SavedObjectsClient': errors, _repository
```

Reply:
Maybe the @elastic/kibana-platform team can help with this type confusion (personally I also find it hard to understand all the differences between `SavedObjectsClient`, `ISavedObjectsRepository` and `SavedObjectsClientContract` :) ).

Reply:
Summary of a Zoom discussion: in the `usageCollector` case we don't have a Kibana request, so we can't use `CoreStart.savedObjects.getScopedClient()` to get a client.

Reply:
Could we try something like this:
```ts
import { SavedObjectsClient } from '../../../core/server'; // exact path TBD

export async function getInternalSavedObjectsClient(core: CoreSetup) {
  return core.getStartServices().then(async ([coreStart]) => {
    const savedObjectsRepo = coreStart.savedObjects.createInternalRepository();
    return new SavedObjectsClient(savedObjectsRepo);
  }
}

Reply:
If something like this works, we preserve the "Saved Objects Client" contract needed for the injected methods, and we still create our own "internal" repo for saved objects which will give us access to all objects in all spaces with this client, rather than it being scoped. This relationship feels more intuitive than using the repo directly, also. (I'm not sure if we have a way to get a scoped version but we've suggested to the Telemetry team that it might be helpful to provide this to the fetch method, similar to the way the ES client is already provided there.)

Reply:
This works in local testing:

```
import { CoreSetup } from 'kibana/server';
// eslint-disable-next-line @kbn/eslint/no-restricted-paths
import { SavedObjectsClient } from '../../../../../src/core/server/saved_objects/service/saved_objects_client';

export async function getInternalSavedObjectsClient(core: CoreSetup) {
  return core.getStartServices().then(async ([coreStart]) => {
    const savedObjectsRepo = coreStart.savedObjects.createInternalRepository();
    return new SavedObjectsClient(savedObjectsRepo);
  });
}
```
(see https://github.com/elastic/kibana/pull/69294/commits/79a1a9a7e035487dbfeb049cf3c9e0df15a386e5 )

Waiting for feedback from @rudolf if it's safe to override the restricted paths error here.

Reply:
Looks great! I _think_ that `kibana/server` is a magic shortcut to that `src/core/server` path so you might be able to just push the `SavedObjectsClient` up into that import with `CoreSetup`, or at least collapse both imports into whichever format you prefer and just have one import line (if I'm not mistaken on this).

Reply:
When importing `SavedObjectsClient` from `kibana/server` instead I get this error during `yarn start` (and yes, I did `clean` and `bootstrap`):

```
  log   [14:38:15.704] [fatal][root] { Error: Cannot find module 'kibana/server'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Module.Hook._require.Module.require (/home/skh/projects/kibana/node_modules/require-in-the-middle/index.js:51:29)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/skh/projects/kibana/x-pack/plugins/ingest_manager/server/collectors/helpers.ts:7:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Module._compile (/home/skh/projects/kibana/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Object.newLoader [as .ts] (/home/skh/projects/kibana/node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module.Hook._require.Module.require (/home/skh/projects/kibana/node_modules/require-in-the-middle/index.js:70:39)
    at Module.Hook._require.Module.require (/home/skh/projects/kibana/node_modules/require-in-the-middle/index.js:70:39)
    at Module.Hook._require.Module.require (/home/skh/projects/kibana/node_modules/require-in-the-middle/index.js:70:39)
    at Module.Hook._require.Module.require (/home/skh/projects/kibana/node_modules/require-in-the-middle/index.js:70:39) code: 'MODULE_NOT_FOUND' }

 FATAL  Error: Cannot find module 'kibana/server'
```
I did not dig deeper, to be honest.